modules.d/98dracut-systemd/module-setup.sh

```diff
@@ -53,4 +53,37 @@ install() {
     inst_simple "$moddir/dracut-tmpfiles.conf" "$tmpfilesdir/dracut-tmpfiles.conf"
 
     inst_multiple sulogin
+
+    if [ -f "$dracutsysrootdir"/etc/crypttab ]; then
+        remove_cryptkeyfiles
```

Comment:
IIRC the problem is key files generated by GRUB at boot time do not exist when dracut builds the initrd, so maybe this logic should be implemented in an emergency hook script (`lib/dracut/hooks/emergency`).

Reply:
The grub generated keyfiles have fixed path and names, so I can predict them.
But modifying the dracut-emergency.sh script when creating the initrd is probably not conform with the usual implementation methods.
I will switch to a hook.

Reply:
Although you can predict the name of the key file, I think GRUB appends the new entry to /etc/crypttab at boot time, but only to the /etc/crypttab file contained in the initrd, not to the system /etc/crypttab, so when dracut builds an initrd, that entry should not be there... but I think we should ask Michael to confirm it :)

Reply:
> But modifying the dracut-emergency.sh script when creating the initrd is probably not conform with the usual implementation methods. I will switch to a hook.

I kept thinking about this and I think I expressed myself badly, just to clarify, the code on the `install` method would be like:

```bash
if [ -z "$keep_cryptkeyfiles" ]; then
    inst_hook emergency 000 "$moddir/remove_cryptkeyfiles.sh"
fi
```

Then, a new `remove_cryptkeyfiles.sh` script would contain the logic of your `remove_cryptkeyfiles` function (you only need to replace `-z "$keep_cryptkeyfiles"` with `-f /etc/crypttab`).

Reply:
> Although you can predict the name of the key file, I think GRUB appends the new entry to /etc/crypttab at boot time, but only to the /etc/crypttab file contained in the initrd, not to the system /etc/crypttab, so when dracut builds an initrd, that entry should not be there... but I think we should ask Michael to confirm it :)

Hm, I was assuming that there can't  be case where the name of the crypt device is not listet in /etc/crypttab of the running system.
I derived this from the code in /etc/grub.d/05_crypttab.
But indeed, I might have misinterpreted the functionality.
@WenhuaChang does grub still write a keyfile for encrypted system root, even if the crypttab in the running system doesn't container an entry for this device?

Reply:
> Hm, I was assuming that there can't be case where the name of the crypt device is not listet in /etc/crypttab of the running system.

Ahh, thanks for the reference, I was missing this part.
Then your code is correct, sorry for the noise.